src/useMultiRootEditor.tsx

```diff
@@ -0,0 +1,524 @@
+
+			return;
+		}
+
+		const editorData = editor.getFullData();
+		const editorAttributes = editor.getRootsAttributes();
+
+		const {
+			addedKeys: newRoots,
+			removedKeys: removedRoots,
+			modifiedKeys: modifiedRoots
+		} = _getStateDiff( editorData, content || {} );
+
+		const {
+			addedKeys: newAttributes,
+			removedKeys: removedAttributes,
+			modifiedKeys: modifiedAttributes
+		} = _getStateDiff( editorAttributes, attributes || {} );
+
+		const newRootAttributes = newAttributes.filter( rootName =>
+			!newRoots.includes( rootName ) && attributes[ rootName ] !== undefined );
+		const removedRootAttributes = removedAttributes.filter( rootName =>
+			!removedRoots.includes( rootName ) && attributes[ rootName ] === undefined );
+		const modifiedRootAttributes = modifiedAttributes.filter( rootName => attributes[ rootName ] !== undefined );
+
+		editor.model.change( writer => {
+			_handleRoots( newRoots, removedRoots, modifiedRoots );
+			_handleRootsAttributes( writer, newRootAttributes, removedRootAttributes, modifiedRootAttributes );
+		} );
+	}, [ content, attributes ] );
+
+	const _getStateDiff = (
+		previousState: Record<string, unknown>,
+		newState: Record<string, unknown>
+	): {
+		addedKeys: Array<string>;
+		removedKeys: Array<string>;
+		modifiedKeys: Array<string>;
+	} => {
+		const previousStateKeys = Object.keys( previousState );
+		const newStateKeys = Object.keys( newState );
+
+		return {
+			addedKeys: newStateKeys.filter( key => !previousStateKeys.includes( key ) ),
+			removedKeys: previousStateKeys.filter( key => !newStateKeys.includes( key ) ),
+			modifiedKeys: previousStateKeys.filter( key => (
+				newStateKeys.includes( key ) &&
+				JSON.stringify( previousState[ key ] ) !== JSON.stringify( newState[ key ] )
+			) )
+		};
+	};
+
+	const _handleRoots = (
+		newRoots: Array<string>,
+		removedRoots: Array<string>,
+		modifiedRoots: Array<string>
+	) => {
+		newRoots.forEach( root => {
+			editor!.addRoot( root, {
+				data: content[ root ] || '',
+				attributes: attributes?.[ root ] || {},
+				isUndoable: true
+			} );
+		} );
+
+		removedRoots.forEach( root => {
+			editor!.detachRoot( root, true );
+		} );
+
+		// If any of the roots content has changed, set the editor data.
+		// Unfortunately, we cannot set the editor data just for one root, so we need to overwrite all roots (`nextProps.data` is an
+		// object with data for each root).
+		if ( modifiedRoots.length ) {
+			editor!.data.set( content, { suppressErrorInCollaboration: true } as any );
+		}
+	};
+
+	const _handleRootsAttributes = (
+		writer: Writer,
+		newRootsAttributes: Array<string>,
+		removedRootsAttributes: Array<string>,
+		modifiedRootsAttributes: Array<string>
+	) => {
+		[ ...newRootsAttributes, ...modifiedRootsAttributes ].forEach( root => {
+			Object.keys( attributes![ root ] ).forEach( attr => {
+				editor!.registerRootAttribute( attr );
+			} );
+
+			writer.clearAttributes( editor!.model.document.getRoot( root )! );
```

Comment:
Can't we simply change the order of handling new&modified attrs with removed attributes instead of preemptively clear all of them?

If not and this line is actually needed to cover some cases, please describe it in a comment.